npm/podController.go

```diff
@@ -0,0 +1,623 @@
+		if err = c.cleanUpDeletedPod(pod); err != nil {
+			metrics.SendErrorLogAndMetric(util.PodID, "Error: %v when pod is in completed state.", err)
+			return fmt.Errorf("Error: %v when when pod is in completed state.\n", err)
+		}
+		return nil
+	}
+
+	err = c.syncAddAndUpdatePod(pod)
+	if err != nil {
+		metrics.SendErrorLogAndMetric(util.PodID, "Error: Failed to sync pod due to %v", err)
+		return fmt.Errorf("Failed to sync pod due to %v\n", err)
+	}
+
+	return nil
+}
+
+func (c *podController) syncAddedPod(podObj *corev1.Pod) error {
+	npmPodObj := newNpmPod(podObj)
+	podNs := util.GetNSNameWithPrefix(podObj.Namespace)
+	podKey, _ := cache.MetaNamespaceKeyFunc(podObj)
+	ipsMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
+	log.Logf("POD CREATING: [%s%s/%s/%s%+v%s]", npmPodObj.PodUID, podNs, npmPodObj.Name, npmPodObj.NodeName, npmPodObj.Labels, npmPodObj.PodIP)
+
+	// Add pod namespace if it doesn't exist
+	var err error
+	if _, exists := c.npMgr.NsMap[podNs]; !exists {
+		// (TODO): need to change newNS function. It always returns "nil"
+		c.npMgr.NsMap[podNs], _ = newNs(podNs)
+		log.Logf("Creating set: %v, hashedSet: %v", podNs, util.GetHashedName(podNs))
+		if err = ipsMgr.CreateSet(podNs, append([]string{util.IpsetNetHashFlag})); err != nil {
+			return fmt.Errorf("[syncAddedPod] Error: creating ipset %s with err: %v", podNs, err)
+		}
+	}
+
+	// Add the pod to its namespace's ipset.
+	log.Logf("Adding pod %s to ipset %s", npmPodObj.PodIP, podNs)
+	if err = ipsMgr.AddToSet(podNs, npmPodObj.PodIP, util.IpsetNetHashFlag, podKey); err != nil {
+		return fmt.Errorf("[syncAddedPod] Error: failed to add pod to namespace ipset with err: %v", err)
+	}
+
+	// Get lists of podLabelKey and podLabelKey + podLavelValue ,and then start adding them to ipsets.
+	addToIPSets := util.GetIPSetListFromLabels(npmPodObj.Labels)
+	for _, addIPSetName := range addToIPSets {
+		log.Logf("Adding pod %s to ipset %s", npmPodObj.PodIP, addIPSetName)
+		if err = ipsMgr.AddToSet(addIPSetName, npmPodObj.PodIP, util.IpsetNetHashFlag, podKey); err != nil {
+			return fmt.Errorf("[syncAddedPod] Error: failed to add pod to label ipset with err: %v", err)
+		}
+	}
+
+	// Add pod's named ports from its ipset.
+	if err = manageNamedPortIpsets(ipsMgr, npmPodObj.ContainerPorts, podKey, npmPodObj.PodIP, AddNamedPortIpsets); err != nil {
+		return fmt.Errorf("[syncAddedPod] Error: failed to add pod to named port ipset with err: %v", err)
+	}
+
+	// add the Pod info to the podMap
+	c.npMgr.PodMap[podKey] = npmPodObj
+	return nil
+}
+
+// syncAddAndUpdatePod handles updating pod ip in its label's ipset.
+func (c *podController) syncAddAndUpdatePod(newPodObj *corev1.Pod) error {
+	log.Logf("[syncAddAndUpdatePod]")
+
+	podKey, _ := cache.MetaNamespaceKeyFunc(newPodObj)
+	newPodObjNs := util.GetNSNameWithPrefix(newPodObj.ObjectMeta.Namespace)
+	ipsMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
+
+	// Add pod namespace if it doesn't exist
+	var err error
```

Comment:
this paradigm will need to change in the future when using NPMError

Reply:
Yes. It will be handled with another PR.